### PR TITLE
[ALLUXIO-3127] Creating/loading an Inode does not load ufs metadata for ancestors when recursive is set to true

### PR DESCRIPTION
**Pull request by bf8086**
https://alluxio.atlassian.net/browse/ALLUXIO-3127

### CODE REVIEW

**Comment by bf8086** [comment]
@gene @aaudiber PTAL.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 1]
```diff
@@ -1008,7 +1010,21 @@ public void syncPersistDirectory(InodeDirectory dir, JournalContext journalConte
...
+              UfsStatus status = ufs.getStatus(ufsUri);
+              if (status.isFile()) {
+                throw new InvalidPathException(String.format(
+                    "Error persisting directory. A file exists at the UFS location %s.", ufsUri));
+              }
+              dir.setOwner(status.getOwner())
+                  .setGroup(status.getGroup())
+                  .setMode(status.getMode());
+            } catch (FileNotFoundException e) {
+              // Ignore exception if file is not found
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: Should you log a warning? Also, should you throw an exception so it is retried?

File: tests/src/test/java/alluxio/master/file/FileSystemMasterIntegrationTest.java [hunk 2]
```diff
@@ -944,6 +946,50 @@ public void unavailableUfsRecursiveCreate() throws Exception {
...
 
+  /**
+   * Tests creating a directory in a nested directory load the UFS status of Inodes on the path.
+   */
+  @Test
+  public void createDirectoryInNestedDirectories() throws Exception {
+    String ufs = Configuration.get(PropertyKey.MASTER_MOUNT_TABLE_ROOT_UFS);
+    FileUtils.createDir(Paths.get(ufs, "d1", "d2", "d3").toString());
+    String parentPath = Paths.get(ufs, "d1").toString();
+    FileUtils.changeLocalFilePermission(parentPath, new Mode((short) 0700).toString());
```
Comment (gpang) [inline tests/src/test/java/alluxio/master/file/FileSystemMasterIntegrationTest.java]: is this mode different from the mode of the final component (d3)?

File: tests/src/test/java/alluxio/master/file/FileSystemMasterIntegrationTest.java [hunk 3]
```diff
@@ -944,6 +946,50 @@ public void unavailableUfsRecursiveCreate() throws Exception {
...
+
+  /**
+   * Tests listing a directory in a nested directory load the UFS status of Inodes on the path.
+   */
+  @Test
+  public void loadMetadataInNestedDirectories() throws Exception {
+    String ufs = Configuration.get(PropertyKey.MASTER_MOUNT_TABLE_ROOT_UFS);
+    FileUtils.createDir(Paths.get(ufs, "d1", "d2", "d3").toString());
+    String parentPath = Paths.get(ufs, "d1").toString();
+    FileUtils.changeLocalFilePermission(parentPath, new Mode((short) 0700).toString());
```
Comment (gpang) [inline tests/src/test/java/alluxio/master/file/FileSystemMasterIntegrationTest.java]: is this mode different from the mode of the final component (d3)?

Comment (bf8086) [reply tests/src/test/java/alluxio/master/file/FileSystemMasterIntegrationTest.java]: Yeah the default permission is 0755.

File: tests/src/test/java/alluxio/master/file/FileSystemMasterIntegrationTest.java (hunk 2 above)
Comment (bf8086) [reply tests/src/test/java/alluxio/master/file/FileSystemMasterIntegrationTest.java]: d3 should use the default permission which is 0755.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java (hunk 1 above)
Comment (bf8086) [reply core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: Good catch. Will do.

File: core/common/src/main/java/alluxio/util/io/PathUtils.java [hunk 4]
```diff
@@ -76,14 +76,14 @@ public static String concatPath(Object base, Object... paths) throws IllegalArgu
     Preconditions.checkArgument(paths != null, "Failed to concatPath: a null set of paths");
     List<String> trimmedPathList = new ArrayList<>();
     String trimmedBase =
-        CharMatcher.is(AlluxioURI.SEPARATOR.charAt(0)).trimTrailingFrom(base.toString().trim());
+        CharMatcher.is(AlluxioURI.SEPARATOR.charAt(0)).trimTrailingFrom(base.toString());
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/util/io/PathUtils.java]: why is this change needed?

File: tests/src/test/java/alluxio/master/file/FileSystemMasterIntegrationTest.java [hunk 5]
```diff
@@ -846,6 +848,8 @@ public void unavailableUfsRecursiveCreate() throws Exception {
         .thenReturn(new UfsDirectoryStatus("test", "owner", "group", (short) 511));
     Mockito.when(mockUfs.mkdirs(Matchers.eq(ufsBase + "/dir1"), Matchers.any()))
         .thenThrow(new IOException("ufs unavailable"));
+    Mockito.when(mockUfs.getStatus(Matchers.any()))
```
Comment (aaudiber) [inline tests/src/test/java/alluxio/master/file/FileSystemMasterIntegrationTest.java]: we might want to mock different file statuses in the future - better to match the exact string that we want to map to `new UfsDirectoryStatus("test", "test", "test", (short) 0755)`

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 6]
```diff
@@ -1008,7 +1010,23 @@ public void syncPersistDirectory(InodeDirectory dir, JournalContext journalConte
           MkdirsOptions mkdirsOptions =
               MkdirsOptions.defaults().setCreateParent(false).setOwner(dir.getOwner())
                   .setGroup(dir.getGroup()).setMode(new Mode(dir.getMode()));
-          ufs.mkdirs(ufsUri, mkdirsOptions);
+          if (!ufs.mkdirs(ufsUri, mkdirsOptions)) {
+            // Directory already exists. Try loading the status from ufs.
```
Comment (aaudiber) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: I think returning false just means the directory couldn't be created, not that it already exists. It could return false due to a permission issue or because the UFS temporarily unavailable

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 7]
```diff
@@ -1008,7 +1010,23 @@ public void syncPersistDirectory(InodeDirectory dir, JournalContext journalConte
           MkdirsOptions mkdirsOptions =
               MkdirsOptions.defaults().setCreateParent(false).setOwner(dir.getOwner())
                   .setGroup(dir.getGroup()).setMode(new Mode(dir.getMode()));
-          ufs.mkdirs(ufsUri, mkdirsOptions);
+          if (!ufs.mkdirs(ufsUri, mkdirsOptions)) {
```
Comment (aaudiber) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: Does making this change here fix the load metadata case? Why does load metadata involve persisting directories to the UFS?

File: core/common/src/main/java/alluxio/util/io/PathUtils.java (hunk 4 above)
Comment (bf8086) [reply core/common/src/main/java/alluxio/util/io/PathUtils.java]: By fixing this bug we now make the master throw error if a directory is not created correctly in UFS. This triggers the mkdir with space tests to fail because we trim the path before creating it on UFS. This line removed the trim so we can actually create directories with leading or trailing spaces on UFS.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java (hunk 6 above)
Comment (bf8086) [reply core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: I think you are right - it's not restricted to directory exist case. 
In general we want to try creating the directory first, and if it does not work, then it is likely already exist in UFS so we will do a  `getStatus` to grab the metadata. if that still fails then we will report error.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java (hunk 7 above)
Comment (bf8086) [reply core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: It works because currently both load metadata and create directory use `INodeTree.createPath` to create the Inodes leading to the target node. For INodes in the path that only exist in UFS we just create the INodes using the owner and modes passed in which should only apply for the target INode and new directories which are created. Given that we call `syncPersistDirectory` on each directory created, we can determine if there is already UFS directory exists and apply directory status to the new Inode.

File: tests/src/test/java/alluxio/master/file/FileSystemMasterIntegrationTest.java (hunk 5 above)
Comment (bf8086) [reply tests/src/test/java/alluxio/master/file/FileSystemMasterIntegrationTest.java]: Done.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java (hunk 6 above)
Comment (bf8086) [reply core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: I updated the comments to be more accurate.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java (hunk 7 above)
Comment (gpang) [reply core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: Well, one interesting scenario is trying to load metadata, but during the process, someone deletes the underlying directories. In that scenario, what is the correct behavior? Should we just fail the "create path"? The current PR will try to mkdir the UFS dir again.

These types of races are always possible, since we do not control the UFS modifications. We will just have to choose a behavior.
